db/buffered_transaction.go

```diff
@@ -0,0 +1,84 @@
+package db
+
+import (
+	"errors"
+)
+
+// BufferedTransaction buffers the updates in the memory to be later flushed to the underlying Transaction
+type BufferedTransaction struct {
+	updates map[string][]byte
+	txn     Transaction
+}
+
+func NewBufferedTransaction(txn Transaction) *BufferedTransaction {
+	return &BufferedTransaction{
+		txn:     txn,
+		updates: make(map[string][]byte),
+	}
+}
+
+// Discard : see db.Transaction.Discard
+func (t *BufferedTransaction) Discard() error {
+	t.updates = nil
+	return t.txn.Discard()
+}
+
+// Commit : see db.Transaction.Commit
+func (t *BufferedTransaction) Commit() error {
+	if err := t.Flush(); err != nil {
+		return err
+	}
+	t.updates = nil
+	return t.txn.Commit()
+}
+
+// Set : see db.Transaction.Set
+func (t *BufferedTransaction) Set(key, val []byte) error {
+	valueCopy := make([]byte, 0, len(val))
+	t.updates[string(key)] = append(valueCopy, val...)
+	return nil
+}
+
+// Delete : see db.Transaction.Delete
+func (t *BufferedTransaction) Delete(key []byte) error {
+	t.updates[string(key)] = nil
+	return nil
+}
+
+// Get : see db.Transaction.Get
+func (t *BufferedTransaction) Get(key []byte, cb func([]byte) error) error {
+	if value, found := t.updates[string(key)]; found {
+		if value == nil {
+			return ErrKeyNotFound
+		}
+		return cb(value)
+	}
+	return t.txn.Get(key, cb)
+}
+
+// Flush applies the pending changes to the underlying Transaction
+func (t *BufferedTransaction) Flush() error {
+	for key, value := range t.updates {
+		keyBytes := []byte(key)
+		if value != nil {
```

Comment:
nit: we can test `value == nil` and swap the branches.

Reply:
Done